tidy3d/plugins/autograd/invdes/filters.py

```diff
@@ -5,28 +5,69 @@
 
 from ..functions import convolve
 from ..types import KernelType, PaddingType
-from ..utilities import make_kernel
+from ..utilities import get_kernel_size_px, make_kernel
+
+
+def _get_kernel_size(
+    radius: Union[float, Tuple[float, ...]],
+    dl: Union[float, Tuple[float, ...]],
+    size_px: Union[int, Tuple[int, ...]],
+) -> Tuple[int, ...]:
+    """Determine the kernel size based on the provided radius, grid spacing, or size in pixels.
+
+    Parameters
+    ----------
+    radius : Union[float, Tuple[float, ...]]
+        The radius of the kernel. Can be a scalar or a tuple.
+    dl : Union[float, Tuple[float, ...]]
+        The grid spacing. Can be a scalar or a tuple.
+    size_px : Union[int, Tuple[int, ...]]
+        The size of the kernel in pixels for each dimension. Can be a scalar or a tuple.
+
+    Returns
+    -------
+    Tuple[int, ...]
+        The size of the kernel in pixels for each dimension.
+
+    Raises
+    ------
+    ValueError
+        If neither 'size_px' nor both 'radius' and 'dl' are provided.
+    """
+    if size_px is not None:
+        return (size_px,) if np.isscalar(size_px) else tuple(size_px)
+    elif radius is not None and dl is not None:
+        kernel_size = get_kernel_size_px(radius=radius, dl=dl)
+        return (kernel_size,) if np.isscalar(kernel_size) else tuple(kernel_size)
+    else:
+        raise ValueError("Either 'size_px' or both 'radius' and 'dl' must be provided.")
```

Comment:
to remove any ambiguity, it might be best to just only handle specific cases?
```
    if size_px is not None and (radius is None and dl is None):
        return (size_px,) if np.isscalar(size_px) else tuple(size_px)
    elif radius is not None and dl is not None and size_px is None:
        kernel_size = get_kernel_size_px(radius=radius, dl=dl)
        return (kernel_size,) if np.isscalar(kernel_size) else tuple(kernel_size)
    else:
        raise ValueError("Either 'size_px' or both 'radius' and 'dl' must be provided.")
```

Reply:
Ah, my thinking was that if a kernel size in pixels is given, then that would take precedence over radius and dl. Maybe it makes sense to do this and warn the user? Or just error?

Reply:
up to you. but just need to clearly define what happens if all 3 are provided perhaps. I might suggest  error or at least warn? seems like untented use.

Reply:
It is clearly defined, no? If `size_px` is given then that takes precedence over everything else. The reason being that if you supply an explicit size in pixels for the filter, then you probably do not want to calculate a different size from other arguments? Maybe it's enough to add this to the docstring?
I thought this was a relatively common pattern, see for example the `period` argument in [np.interp](https://numpy.org/doc/stable/reference/generated/numpy.interp.html).

Reply:
It is clearly defined, no? If `size_px` is given then that takes precedence over everything else. The reason being that if you supply an explicit size in pixels for the filter, then you probably do not want to calculate a different size from other arguments? Maybe it's enough to add this to the docstring?
I thought this was a relatively common pattern, see for example the `period` argument in [np.interp](https://numpy.org/doc/stable/reference/generated/numpy.interp.html).

Reply:
ok that's fine with me. I tend to be a bit conservative about things like this (tend towards erroring), if only to make things simpler, but if it's ok for numpy then I'm fine with it.